### ISSUE TITLE
Nix flake support

### ISSUE THREAD

**Issue by loafofpiecrust** [terranix/terranix#9]
I use flakes in my personal system config and for various projects. They generally work quite well, better than normal `default.nix` or `shell.nix`. I am setting up a new project and want to try `terranix` but on first glance it doesn't work with flakes. I get an error when I try to run `nix build`.

Here is my flake:
```nix
{
  inputs = { pkgs.url = "github:nixos/nixpkgs/nixos-21.05"; };
  outputs = { self, pkgs }: {
    defaultPackage.x86_64-linux = with import pkgs { system = "x86_64-linux"; };
      stdenv.mkDerivation {
        name = "example";
        src = self;
        buildInputs = [ terranix terraform jq nix ];
        buildPhase = ''
          terranix | jq '.' > config.tf.json
          terraform init
          terraform plan
        '';
      };
  };
}
```

And here is the error:
```
       last 8 log lines:
       > unpacking sources
       > unpacking source archive /nix/store/a0rqhq7wk3r6i3bhc1qp9119w7bb3245-source
       > source root is source
       > patching sources
       > configuring
       > no configure script, doing nothing
       > building
       > error: creating directory '/nix/var': Permission denied
       For full logs, run 'nix log /nix/store/qaqsd36krcs09hxry30h7qgvk5ivk18f-dailp.drv'.
```

I think it originates from `terranix` using `nix-build`, the older system from before flakes, within a flake. Do you think there's an easy fix to this?

**Comment by Mic92** [comment]
So what happens here is that terranix calles nix, which does not work very well while being in the sandbox. If you have a short look at the cli, you will see that you can also call to underlying nix code directly: https://github.com/mrVanDalo/terranix/blob/706b76a1109326931504259d7319ffdf88c2cfc1/bin/terranix#L82 this should give you the json directly and you should be able to than use it for the further terraform steps.

**Comment by mrVanDalo** [comment]
Thanks @Mic92 for the answer. I have no experience with flakes so far, but I plan to dig into it this year, and make terranix easier to use with flakes. @loafofpiecrust [terranix is available in the 21.05 stable channel](https://search.nixos.org/packages?channel=21.05&query=terranix), so maybe no need for flakes, for now?

**Comment by loafofpiecrust** [comment]
Well, I want to use a flake whether or not `terranix` is in `nixpkgs`, because it lets me lock the version of `nixpkgs` that I'm using per project. In the example `flake.nix` I gave, you can see I'm just using `terranix` straight from `nixpkgs`.

I followed @Mic92's advice and adjusted my flake to directly use the terranix `.nix` files. However, then I ran into other issues with my approach because terraform needs internet to download providers. I tried using nixpkgs provider packages, but then the AWS provider still needed network access. So, I was a bit misguided in thinking that I would be able to run `terraform plan` in a sandbox. But using terranix directly rather than the bash script works great, as you can see in my revised flake:

```nix
{
  inputs = { pkgs.url = "github:nixos/nixpkgs/nixos-21.05"; };
  outputs = { self, pkgs }: {
    defaultPackage.x86_64-linux = with import pkgs { system = "x86_64-linux"; };
      let
        terraformConfig = (let
          tf = import "${terranix}/core/default.nix" {
            inherit pkgs;
            terranix_config = { imports = [ ./config.nix ]; };
            strip_nulls = true;
          };
        in writeTextFile {
          name = "terraform-config";
          text = builtins.toJSON (tf.config);
          executable = false;
          destination = "/config.tf.json";
        });
      in stdenv.mkDerivation {
        name = "example";
        src = self;
        installPhase = ''
          mkdir -p $out
          cp -f ${terraformConfig}/config.tf.json $out/
        '';
      };
  };
}
```

There might be a simpler way to get the json file in the right place at the end, but now running `nix build` gives me a nice `./result/config.tf.json` and I can rename `config.nix` to whatever I want! I would consider adding an example like this to the documentation.

**Comment by Mic92** [comment]
Maybe having a helper to cut down on the boiler code would be helpful?

**Comment by mrVanDalo** [comment]
@loafofpiecrust I merged your the #10 and try to make use of it as dev-shell but I haven't figured out how to make properly use of it. `config.tf.json` always stays in the nix-store, which makes it impossible to run terraform in the `result`, because it can't create a state file in the nix-store.
I also have to write some documentation on how to use this flake, and I have a hard time to find a useful example which actually makes sense. I mean yes. I'm able to create the `config.tf.json` with flakes, but that is not a use case isn't it? I at least want to create a command that I can just fire, right?

**Comment by terlar** [comment]
Another alternative to having code in a write-able place is to use the environment variable [TF_DATA_DIR](https://www.terraform.io/docs/cli/config/environment-variables.html#tf_data_dir)

I usually create a derivation that generates TF code inside the nix store and then you can have a shell that enters this directory as deployment context. That way I can ensure what is in the derivation is what is being deployed.

**Comment by mrVanDalo** [comment]
thx @terlar for the input, but it still is not working.

```
[nix-shell:~/dev/terranix/terranix/examples/flake/result]$ ll
total 4.0K
-r--r--r-- 1 root root 345 Jan  1  1970 config.tf.json

[nix-shell:~/dev/terranix/terranix/examples/flake/result]$ echo $TF_DATA_DIR 
/home/palo/foo

[nix-shell:~/dev/terranix/terranix/examples/flake/result]$ terraform init

Initializing the backend...

Initializing provider plugins...
- Finding latest version of hashicorp/local...
- Installing hashicorp/local v2.1.0...
- Installed hashicorp/local v2.1.0 (signed by HashiCorp)

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.
╷
│ Error: Failed to update dependency lock file
│ 
│ Error while writing new dependency lock information to .terraform.lock.hcl: cannot create temporary file to update .terraform.lock.hcl: open .terraform.lock.hcl406915942: read-only file system.
╵
```

**Comment by terlar** [comment]
Hmm, I guess I haven't run in to that, either because I manage the plugins via a pre-prepared terraform including all the plugins I will use or B, using a terraform remote state on S3. In this case it looks related to the plugins though.

For example:
```
terraformFull = final.terraform_0_13.withPlugins (p: [ p.aws p.datadog p.null ]);
```

What about if you set [TF_PLUGIN_CACHE_DIR](https://www.terraform.io/docs/cli/config/config-file.html#provider-plugin-cache)?

Or perhaps since, I'm still using 0.13 it is something that has changed further since then.

**Comment by mrVanDalo** [comment]
I'm using terraform 0.15. Thanks for the quick response @terlar I will offer this function anyway, but write in the documentation, that it's not that easy to use, but a bonus at least would be, flakes can be used instead of `pkgs.fetchGit` or something. That is at least a plus. I'm also thinking of creating a package that can be called, and prints out the json. So you would have a pipeline like
```
terraform_config > config.tf.json && terraform
```
But I'm experimenting what could be a good setup.

**Comment by terlar** [comment]
Seems there are a lot of issues related to this lock file. Hopefully they will add an option to disable it completely, please cast your votes:
https://github.com/hashicorp/terraform/issues/27241

**Comment by mrVanDalo** [comment]
@terlar @loafofpiecrust @l5r I merged #10 and cleaned it a bit. Thanks for your input. If you have any further ideas, please let me know.
* https://github.com/terranix/terranix/tree/develop/examples/flake
* https://github.com/terranix/terranix/blob/develop/flake.nix
I'm closing this issue, feel free to reopen it again.
